<a href="https://colab.research.google.com/github/cristiandarioortegayubro/UNI/blob/main/colab/pd_02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![](https://github.com/cristiandarioortegayubro/UNI/blob/main/UNI%2039%20Aniversario.jpg?raw=true)

![pandas](https://upload.wikimedia.org/wikipedia/commons/thumb/e/ed/Pandas_logo.svg/250px-Pandas_logo.svg.png)

**En este cuaderno continuamos con el desarrollo de los temás más usuales y utilizados del módulo Pandas...**

https://pandas.pydata.org/docs/user_guide/index.html

***Las estructuras de datos de Pandas nos proporcionan un formato para organizar, administrar y almacenar datos. El conocimiento de estas estructuras de datos (series, dataframe y paneles de datos) resultará útil cuando se busque solucionar problemas o realizar distintas operaciones en los datos. Las estructuras de datos de Pandas son diferentes de las estructuras de datos de Python (tuplas, listas y diccionarios) por una simple razón: las estructuras de datos de Pandas fueron creadas para tareas de análisis de datos.***

# **Carga de bibliotecas necesarias**

In [ ]:
import pandas as pd
import numpy as np

# **Obtención de datos**

In [ ]:
datos = "https://raw.githubusercontent.com/cristiandarioortegayubro/UNI/main/analisis_facturas_ventas.csv"

In [ ]:
df = pd.read_csv(datos)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 347 entries, 0 to 346
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   DOCUMENTO              347 non-null    object
 1   ORGANIZACION           347 non-null    object
 2   FECHA                  347 non-null    object
 3   TIPOFACTURA            347 non-null    object
 4   PRODUCTO               347 non-null    object
 5   CANTIDAD               347 non-null    int64 
 6   PRECIO                 347 non-null    int64 
 7   IMPORTEMONTRANSACCION  347 non-null    int64 
 8   MONEDA                 347 non-null    object
dtypes: int64(3), object(6)
memory usage: 24.5+ KB


***Podemos observar que, como el tipo de datos de la fecha no es el que corresponde...***

In [ ]:
df["FECHA"] = pd.to_datetime(df.FECHA, infer_datetime_format=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 347 entries, 0 to 346
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   DOCUMENTO              347 non-null    object        
 1   ORGANIZACION           347 non-null    object        
 2   FECHA                  347 non-null    datetime64[ns]
 3   TIPOFACTURA            347 non-null    object        
 4   PRODUCTO               347 non-null    object        
 5   CANTIDAD               347 non-null    int64         
 6   PRECIO                 347 non-null    int64         
 7   IMPORTEMONTRANSACCION  347 non-null    int64         
 8   MONEDA                 347 non-null    object        
dtypes: datetime64[ns](1), int64(3), object(5)
memory usage: 24.5+ KB


*Nuestro DataFrame contiene 347 registros y 9 columnas...*

# **Selección de datos...**

## **Selección por etiquetas**

***Pandas proporciona un conjunto de métodos para tener una indexación basada en etiquetas. Este es un protocolo estricto basado en la inclusión. Cada etiqueta solicitada debe existir, tanto en el index como en las columnas. Al seleccionar por ejemplo filas (index) se incluye el límite inicial indicado y el límite final. Los números enteros son etiquetas válidas, pero se refieren a las etiquetas del index (cuando la etiqueta es numérica) y no a la posición de la fila.***

**En los parámetros de la selección, siempre se indica primero las filas y luego las columnas.. En el caso de seleccionar de un DataFrame, se hace de la siguiente forma:**

~~~python
df.loc[fila inicial : fila final, columna inicial : columna final]
~~~


***Seleccionando tres filas, desde la fila 0 hasta la fila 2...***

In [ ]:
df.loc[0:2]

DOCUMENTO      ORGANIZACION  ... IMPORTEMONTRANSACCION            MONEDA
0  B-0004-00000007  Consumidor Final  ...            1300000000  Pesos Argentinos
1  A-0004-00000039      Brillamás SA  ...             380000000  Pesos Argentinos
2  A-0004-00000039      Brillamás SA  ...             260000000  Pesos Argentinos

[3 rows x 9 columns]

***Seleccionando cuatro filas, desde la fila 2 hasta la fila 5...***

In [ ]:
df.loc[2:5]

DOCUMENTO    ORGANIZACION  ... IMPORTEMONTRANSACCION            MONEDA
2  A-0004-00000039    Brillamás SA  ...             260000000  Pesos Argentinos
3  A-0004-00000038  Horizonte S.A.  ...            1200000000  Pesos Argentinos
4  A-0004-00000037   Galardon S.A.  ...             960000000  Pesos Argentinos
5  A-0004-00000036   Galardon S.A.  ...            4000000000  Pesos Argentinos

[4 rows x 9 columns]

In [ ]:
df.loc[2:5,]

DOCUMENTO    ORGANIZACION  ... IMPORTEMONTRANSACCION            MONEDA
2  A-0004-00000039    Brillamás SA  ...             260000000  Pesos Argentinos
3  A-0004-00000038  Horizonte S.A.  ...            1200000000  Pesos Argentinos
4  A-0004-00000037   Galardon S.A.  ...             960000000  Pesos Argentinos
5  A-0004-00000036   Galardon S.A.  ...            4000000000  Pesos Argentinos

[4 rows x 9 columns]

In [ ]:
df.loc[2:5,:]

DOCUMENTO    ORGANIZACION  ... IMPORTEMONTRANSACCION            MONEDA
2  A-0004-00000039    Brillamás SA  ...             260000000  Pesos Argentinos
3  A-0004-00000038  Horizonte S.A.  ...            1200000000  Pesos Argentinos
4  A-0004-00000037   Galardon S.A.  ...             960000000  Pesos Argentinos
5  A-0004-00000036   Galardon S.A.  ...            4000000000  Pesos Argentinos

[4 rows x 9 columns]

In [ ]:
df.loc[2:5,"DOCUMENTO":"CANTIDAD"]

DOCUMENTO    ORGANIZACION  ...                 PRODUCTO CANTIDAD
2  A-0004-00000039    Brillamás SA  ...                    Mouse   200000
3  A-0004-00000038  Horizonte S.A.  ...  Impresora Multifuncion    100000
4  A-0004-00000037   Galardon S.A.  ...  Impresora Multifuncion     80000
5  A-0004-00000036   Galardon S.A.  ...                       PC    40000

[4 rows x 6 columns]

***Seleccionando dos columnas, TIPOFACTURA y FECHA...***

In [ ]:
df.loc[:,["TIPOFACTURA","FECHA"]]

TIPOFACTURA      FECHA
0       Factura 2022-01-25
1       Factura 2022-01-19
2       Factura 2022-01-19
3       Factura 2022-01-11
4       Factura 2022-01-07
..          ...        ...
342     Factura 2017-01-06
343     Factura 2017-01-06
344     Factura 2017-01-05
345     Factura 2017-01-03
346     Factura 2017-01-02

[347 rows x 2 columns]

**Seleccionando filas y columnas**

In [ ]:
df.loc[7:9,["FECHA","TIPOFACTURA"]]

FECHA TIPOFACTURA
7 2021-12-17     Factura
8 2021-12-15     Factura
9 2021-12-15     Factura

In [ ]:
df.loc[[7,9],["FECHA","TIPOFACTURA"]]

FECHA TIPOFACTURA
7 2021-12-17     Factura
9 2021-12-15     Factura

In [ ]:
df.loc[[7,9,8],["FECHA","TIPOFACTURA"]]

FECHA TIPOFACTURA
7 2021-12-17     Factura
9 2021-12-15     Factura
8 2021-12-15     Factura

In [ ]:
df.loc[:,["FECHA","TIPOFACTURA"]].head()

FECHA TIPOFACTURA
0 2022-01-25     Factura
1 2022-01-19     Factura
2 2022-01-19     Factura
3 2022-01-11     Factura
4 2022-01-07     Factura

## **Selección por posición**

***Pandas proporciona un conjunto de métodos para obtener una indexación puramente basada en enteros. La semántica sigue de cerca lo definido en Python y NumPy. Estos son indexación. Al cortar, se incluye el límite inicial $n$, mientras que se excluye el límite final $m$, es decir desde $n$ hasta $m-1$***

**En los parámetros de la selección, siempre se indica primero las filas y luego las columnas. En el caso de seleccionar de un DataFrame, se hace de la siguiente forma:**

~~~python
df.loc[fila posición inicial : fila posición final, columna posición inicial : columna posición final]
~~~

**Selección por ubicación del dato**

In [ ]:
df.iloc[6]

DOCUMENTO                    A-0004-00000036
ORGANIZACION                   Galardon S.A.
FECHA                    2022-01-05 00:00:00
TIPOFACTURA                          Factura
PRODUCTO                        Modem Router
CANTIDAD                              100000
PRECIO                                  2100
IMPORTEMONTRANSACCION              210000000
MONEDA                      Pesos Argentinos
Name: 6, dtype: object

In [ ]:
df.iloc[3:5,0:2]

DOCUMENTO    ORGANIZACION
3  A-0004-00000038  Horizonte S.A.
4  A-0004-00000037   Galardon S.A.

***También se pueden presentar las posiciones de las filas y de las columnas como una lista...***

In [ ]:
df.iloc[[0,2,4],[0,2,3]]

DOCUMENTO      FECHA TIPOFACTURA
0  B-0004-00000007 2022-01-25     Factura
2  A-0004-00000039 2022-01-19     Factura
4  A-0004-00000037 2022-01-07     Factura

In [ ]:
df.iloc[[1,2,4],[0,2]]

DOCUMENTO      FECHA
1  A-0004-00000039 2022-01-19
2  A-0004-00000039 2022-01-19
4  A-0004-00000037 2022-01-07

## **Encadenamiento**

**Con esta forma de seleccionar datos, primero se deben elegir los nombres de las columnas deseadas y luego el rango de filas a mostrar...**

In [ ]:
df[["DOCUMENTO", "FECHA"]][0:4]

DOCUMENTO      FECHA
0  B-0004-00000007 2022-01-25
1  A-0004-00000039 2022-01-19
2  A-0004-00000039 2022-01-19
3  A-0004-00000038 2022-01-11

In [ ]:
df[["FECHA", "PRODUCTO", "PRECIO", "DOCUMENTO"]][0:10]

FECHA                 PRODUCTO  PRECIO        DOCUMENTO
0 2022-01-25                       PC  130000  B-0004-00000007
1 2022-01-19                  Teclado    1900  A-0004-00000039
2 2022-01-19                    Mouse    1300  A-0004-00000039
3 2022-01-11  Impresora Multifuncion    12000  A-0004-00000038
4 2022-01-07  Impresora Multifuncion    12000  A-0004-00000037
5 2022-01-05                       PC  100000  A-0004-00000036
6 2022-01-05             Modem Router    2100  A-0004-00000036
7 2021-12-17                       PC   90909  B-0004-00000006
8 2021-12-15    Mercaderia de Reventa    3100  A-0004-00000035
9 2021-12-15     Productos Terminados    1500  A-0004-00000035

## **Filtros**

**Podemos llevar a cabo algunas opciones para filtrar nuestros datos, incluidas las máscaras booleanas y algunos métodos especiales. Con las máscaras booleanas, probamos nuestros datos contra algún valor y obtenemos una estructura de la misma forma, excepto que está llena de valores que son True o False.**

In [ ]:
df.head(3)

DOCUMENTO      ORGANIZACION  ... IMPORTEMONTRANSACCION            MONEDA
0  B-0004-00000007  Consumidor Final  ...            1300000000  Pesos Argentinos
1  A-0004-00000039      Brillamás SA  ...             380000000  Pesos Argentinos
2  A-0004-00000039      Brillamás SA  ...             260000000  Pesos Argentinos

[3 rows x 9 columns]

**Filtrando los datos para valores de PRECIO mayores a $100000$, en este caso, se muestra la serie PRECIO con datos booleanos que cumplen la condición...**

In [ ]:
df.PRECIO > 100000

0       True
1      False
2      False
3      False
4      False
       ...  
342    False
343    False
344    False
345    False
346    False
Name: PRECIO, Length: 347, dtype: bool

**Si lo que se quiere es ver el resultante de la condición en todo el DataFrame, entonces...**

In [ ]:
df[df.PRECIO > 100000]

DOCUMENTO      ORGANIZACION  ... IMPORTEMONTRANSACCION            MONEDA
0    B-0004-00000007  Consumidor Final  ...            1300000000  Pesos Argentinos
13   B-0004-00000005  Consumidor Final  ...            1239669400  Pesos Argentinos
61   A-0001-00000247   Donadelli S.R.L  ...            1500000000  Pesos Argentinos
71   A-0003-00000001  Manganello S.R.L  ...            2200000000  Pesos Argentinos
100  A-0001-00000220    Hernandez S.A.  ...            1050600000  Pesos Argentinos
139  A-0001-00000194    Horizonte S.A.  ...            1500000000  Pesos Argentinos
141  A-0001-00000192  Manganello S.R.L  ...            1055000000  Pesos Argentinos
157  A-0001-00000179       Jameson SRL  ...            1659800000  Pesos Argentinos

[8 rows x 9 columns]

**Para visualizar más de dos condiciones verdaderas, es decir todas las condiciones verdaderas se usa el símbolo &...**

In [ ]:
df.loc[(df.CANTIDAD > 10000) & (df.PRECIO > 100000)]

DOCUMENTO      ORGANIZACION  ... IMPORTEMONTRANSACCION            MONEDA
71  A-0003-00000001  Manganello S.R.L  ...            2200000000  Pesos Argentinos

[1 rows x 9 columns]

**Para visualizar más de dos condiciones y alguna de ellas es verdadera, se usa el símbolo |**

In [ ]:
df.loc[(df.CANTIDAD > 20000) | (df.PRECIO > 100000)]

DOCUMENTO  ...            MONEDA
0    B-0004-00000007  ...  Pesos Argentinos
1    A-0004-00000039  ...  Pesos Argentinos
2    A-0004-00000039  ...  Pesos Argentinos
3    A-0004-00000038  ...  Pesos Argentinos
4    A-0004-00000037  ...  Pesos Argentinos
..               ...  ...               ...
335  B-0001-00000050  ...  Pesos Argentinos
337  B-0001-00000049  ...  Pesos Argentinos
341  B-0001-00000047  ...  Pesos Argentinos
342  B-0001-00000047  ...  Pesos Argentinos
343  B-0001-00000047  ...  Pesos Argentinos

[183 rows x 9 columns]

**Para un conjunto de valores entre un intervalo determinado...**

In [ ]:
df.loc[df.PRECIO.between(130000, 180000)]

DOCUMENTO      ORGANIZACION  ... IMPORTEMONTRANSACCION            MONEDA
0    B-0004-00000007  Consumidor Final  ...            1300000000  Pesos Argentinos
61   A-0001-00000247   Donadelli S.R.L  ...            1500000000  Pesos Argentinos
139  A-0001-00000194    Horizonte S.A.  ...            1500000000  Pesos Argentinos
157  A-0001-00000179       Jameson SRL  ...            1659800000  Pesos Argentinos

[4 rows x 9 columns]

In [ ]:
df.loc[df.PRECIO.between(100000, 180000) & df.CANTIDAD.between(20000, 30000)]

DOCUMENTO      ORGANIZACION  ... IMPORTEMONTRANSACCION            MONEDA
11  A-0004-00000033      Brillamás SA  ...            3000000000  Pesos Argentinos
71  A-0003-00000001  Manganello S.R.L  ...            2200000000  Pesos Argentinos

[2 rows x 9 columns]

In [ ]:
df.loc[df.PRECIO.between(100000, 180000) | df.CANTIDAD.between(20000, 30000)]

DOCUMENTO  ...            MONEDA
0    B-0004-00000007  ...  Pesos Argentinos
5    A-0004-00000036  ...  Pesos Argentinos
11   A-0004-00000033  ...  Pesos Argentinos
13   B-0004-00000005  ...  Pesos Argentinos
54   A-0004-00000006  ...  Pesos Argentinos
..               ...  ...               ...
335  B-0001-00000050  ...  Pesos Argentinos
337  B-0001-00000049  ...  Pesos Argentinos
338  B-0001-00000048  ...  Pesos Argentinos
341  B-0001-00000047  ...  Pesos Argentinos
343  B-0001-00000047  ...  Pesos Argentinos

[61 rows x 9 columns]

# **Sumando y eliminando datos**

**Muy rara vez se obtendrán datos donde no se quiera agregar ni eliminar algo.**

In [ ]:
df.head(3)

DOCUMENTO      ORGANIZACION  ... IMPORTEMONTRANSACCION            MONEDA
0  B-0004-00000007  Consumidor Final  ...            1300000000  Pesos Argentinos
1  A-0004-00000039      Brillamás SA  ...             380000000  Pesos Argentinos
2  A-0004-00000039      Brillamás SA  ...             260000000  Pesos Argentinos

[3 rows x 9 columns]

## **Creando columnas**

**Se puede agregar una columna al final de las existentes en el DataFrame, de la siguiente manera...**

In [ ]:
df["PROVINCIA"] = "Mendoza"
df.head(3)

DOCUMENTO      ORGANIZACION  ...            MONEDA PROVINCIA
0  B-0004-00000007  Consumidor Final  ...  Pesos Argentinos   Mendoza
1  A-0004-00000039      Brillamás SA  ...  Pesos Argentinos   Mendoza
2  A-0004-00000039      Brillamás SA  ...  Pesos Argentinos   Mendoza

[3 rows x 10 columns]

In [ ]:
df["NACIONAL"] = df.MONEDA == "Pesos Argentinos"
df.head(3)

DOCUMENTO      ORGANIZACION  ... PROVINCIA NACIONAL
0  B-0004-00000007  Consumidor Final  ...   Mendoza     True
1  A-0004-00000039      Brillamás SA  ...   Mendoza     True
2  A-0004-00000039      Brillamás SA  ...   Mendoza     True

[3 rows x 11 columns]

**También se puede seleccionar la ubicación de la columna a crear, de la siguiente forma...**

In [ ]:
df.insert(7, column="TRANSACCION", value=df.CANTIDAD * df.PRECIO)

In [ ]:
df.head(3)

DOCUMENTO      ORGANIZACION  ... PROVINCIA NACIONAL
0  B-0004-00000007  Consumidor Final  ...   Mendoza     True
1  A-0004-00000039      Brillamás SA  ...   Mendoza     True
2  A-0004-00000039      Brillamás SA  ...   Mendoza     True

[3 rows x 12 columns]

## **Eliminando columnas**

In [ ]:
df.drop(columns="IMPORTEMONTRANSACCION", inplace=True)

In [ ]:
df.head(3)

DOCUMENTO      ORGANIZACION  ... PROVINCIA NACIONAL
0  B-0004-00000007  Consumidor Final  ...   Mendoza     True
1  A-0004-00000039      Brillamás SA  ...   Mendoza     True
2  A-0004-00000039      Brillamás SA  ...   Mendoza     True

[3 rows x 11 columns]

**Hemos llegado al fin de este cuaderno, felicitaciones seguimos con Pandas, ahora a seguir en el proximo cuaderno...** 

![nenunenu](https://emojipedia-us.s3.dualstack.us-west-1.amazonaws.com/thumbs/60/apple/124/smiling-face-with-smiling-eyes_1f60a.png)